### ISSUE TITLE
Q: Smallest unit of work in an RSpec suites

### ISSUE THREAD

**Issue by agis** [Shopify/ci-queue#129]
Apologies if this is not the place to ask questions, but I couldn't find a better one ([Discussions](https://github.com/zeit/next.js/discussions) would be a great fit here).

I have a few questions regarding how ci-queue splits tests in an RSpec test suite.

1. is the smallest unit of work an RSpec example (i.e. an `it` block)? In other words, are individual items in the queue just specific examples? Or maybe is it whole example groups (i.e. `describe` with multiple `it` blocks inside)?
2. if so, I guess that you've found it to be more optimal as opposed to having whole spec files be the smallest unit of work. Is that so? In particular, I was thinking that scheduling individual examples might perform worse than scheduling a whole spec file. For example, consider the following scenario:
```
describe "foo" do
   # a very expensive factory
  let!(:car) { Factory.create(:car) }

  it "works" do
    # some tests...
  end

  it "moves" do
   # some tests...
  end
end
```

2. (cont.) In that case above, I assume that, due to Rails' transactional tests, scheduling the two `it` blocks in different workers might be less optimal than scheduling the whole `describe` block (or whole file) in a single worker. That's because in the latter case, the expensive factory would be created once, while in the former case it'll be created two times (one in each worker).

In general, I'm interested in any insights you might have regarding the optimal way to split a large RSpec suite, with regards to the "smallest unit of work".

Thanks in advance!

**Comment by casperisfine** [comment]
> 1. the smallest unit of work an RSpec example (i.e. an `it` block)?

Yes.

> if so, I guess that you've found it to be more optimal as opposed to having whole spec files be the smallest unit of work. Is that so?

Exactly. The smallest the better. It's common to see giant `_spec.rb` files that takes minutes. It's way too big of a unit of work, and it force weird behaviors such as splitting big test files in smaller ones.

>  I was thinking that scheduling individual examples might perform worse than scheduling a whole spec file. For example, consider the following scenario

That is one downside indeed. But it's acknowledged as as tradeoff. That's why `before(:all)` and friends is discouraged. If you want to parallelize your workload efficiently, you have to give up most of this statefulness.

In theory you could try to be smart, and schedule unit of works that depends on these same setup blocks on the same workers, but it would be brittle and fragile.

**Comment by agis** [comment]
Thanks for clarifying, @casperisfine!

Regarding `before(:all)`, is it actually discouraged or completely disallowed? From the readme states that it's disallowed. If so, is that because of some technical limitation or it's just considered an anti-pattern (for performance) by ci-queue?

Side question: we're a big Rails 5.2 app with transactional specs, but the codebase is far from being threadsafe. We currently run in an in-house CI system, using containers as workers that execute specs (each container runs a MySQL instance, an ElasticSearch instance, redis etc.) Now, I'm considering using ci-queue instead of rolling a revamped solution again (the current one has some limitations). Does anything from the aforementioned setup sounds like it wouldn't work with ci-queue?

Side question #2: do you have any numbers you can share regarding your Ruby test suite timings and resources? In particular I'd be interested in no. of workers, no. of specs (examples or files) total test execution time with ci-queue, and whether you use in-house workers or something like CircleCI.

Thanks again!

**Comment by casperisfine** [comment]
> Regarding `before(:all)`, is it actually discouraged or completely disallowed?

It's been a while I implemented that, so not 100% sure. But IIRC the goal was to entirely disallow it, but there's some ways to use it anyway because I couldn't find a proper way to entirely disallow it.

What's certain is that it don't work properly with `rspec-queue`, so getting rid of all occurrences is usually the first step to convert a test suite to `ci-queue`.

> is that because of some technical limitation or it's just considered an anti-pattern (for performance) by ci-queue?

Kinda both. The RSpec core team kinda regrets adding it, and I couldn't figure out how to properly add support for it in `rspec-queue`.

It's also an anti-pattern from a `ci-queue` perspective, because it's easily source of state leak.

I think it's theoretically possible to support it, I just couldn't find a way.

> Does anything from the aforementioned setup sounds like it wouldn't work with ci-queue?

Nothing comes to mind. Thread safety isn't a concern at all. 

The only thing to worry about is that the rspec integration has rough edges (such as the `before(:all)` limitation). It's note quite on par with the minitest one.

> do you have any numbers you can share regarding your Ruby test suite timings and resources? 

In short, our main CI uses ci-queue for:
  - 3 test suites (unit, slow, browser)
  - About 275 concurrent ci-queue workers
  - Over 150k unit tests
  - For around 36 hours of aggregated run-time (excluding setup time, test runtime only), which take between 15 and 20m realtime.

All this constantly change though.

**Comment by agis** [comment]
Thanks for following up, really appreciated!

Some last questions (I promise):

- How is parallel execution achieved? Is each worker an isolated container? Or do you parallelize tests by using something like [parallel_tests](https://github.com/grosser/parallel_tests) (meaning some multi-process/threaded solution) inside the same host
- do workers use the same database/storage instances? Or does each worker have its own database?
- do you allocate 1 CPU core per worker? or more?
- do you use RSpec, minitest, or both?
- do the 3 test suites (unit, slow, browser) run in parallel? Or they're distinct steps of a pipeline that are executed sequentially?
- when a failure is reported in an example, is it reported to developers immediately, or do they have to wait for the whole build to complete?

Thanks a lot! :)

**Comment by casperisfine** [comment]
> * How is parallel execution achieved? Is each worker an isolated container?

Yeah, each worker is spawned by a Buildkite job.

> do workers use the same database/storage instances?

No, workers share nothing (beside the redis queue obviously).

Sharing datastore would entirely defeat the purpose.

> do you allocate 1 CPU core per worker? or more?

It's entirely dependent on workload.

> do you use RSpec, minitest, or both?

Mostly minitest, but some projects use rspec.

> do the 3 test suites (unit, slow, browser) run in parallel?

All in parallel.

> when a failure is reported in an example, is it reported to developers immediately, or do they have to wait for the whole build to complete?

Kinda both but not really.

The error is immediately printed by the job that ran it, but you have to wait for everything to complete to see the full summary.

Stopping on the first error is possible, but the reasoning is that once you are on CI, you paid for setup time etc already, so might as well give the developer as much feedback as possible (the entire list of broken test).

**Comment by agis** [comment]
Thanks a ton for all the clarifications, they really helped! :pray:
 
We'll probably give ci-queue a try, sometime in the future. As for now, I'm closing this issue since all questions have been answered.

Thanks :)
